### PR TITLE
`forall` quantifier and mandatory declarations for type variables in new analysis

### LINKED ISSUE #14570: Distinguish fixed free types from generic type variables in experimental solidity.
Currently, all type variables, whether implicitly assigned or explicitly mentioned in the sources, are treated as generic type variables (``TypeVariable``). It is worthwhile to distinguish those from contextually fixed free types. (See also "Distinction between Fixed Free Type Variables and Schematic Type Variables" in https://notes.ethereum.org/_OSmtx9aQAOHQXwa60IDsQ for some previous notes on that)

The unification behaviour of fixed free type variables is simple: they only unify with themselves (including identical sort constraints), nothing else.

The assignment between a name (like ``a`` in ``function f(x:a)``) and a free type variable is scoped (here valid for the function signature and body - for type variables in type classes for the entire type class definition), but for safety their internal representation in the type system should be globally unique.

One of the main advantages of having free type variables is that they can prevent accidental, undesired unification.

The process of type-checking a function should involve a post-processing step on the type, which replaces all generic type variables by free types. That prevents further unification of the function type after its passed type inference (unifying the actual function signature later on is erroneous).
Conversely, that means that *importing* a function (when referred to with an identifier) into local context will involve the opposite transformation, *generalization*, i.e. replacing the fixed free types with fresh type variables (the analog to this currently happens using ``polymorphicInstance``, but nothing prevents us from forgetting to use this function and import with not-freshened type variables, which will result in invalid behaviour - if we have assign fixed free types to the function instead, importing that type without generalization will usually result in unification errors rather than invalid behaviour)

Note that this approach is expected to break mutually recursive functions for now (they're not expected to properly work in the current version anyways, though) - that will be remedied after we can order type inference using the upcoming function call graph (https://github.com/ethereum/solidity/issues/14532), but there's no reason to delay this issue until that's done.

Furthermore, explicitly mentioned type variables should be assigned fixed free types to begin with instead of generic type variables. Note that reference resolution currently also doesn't scope explicitly mentioned type variables correctly, so don't be surprised by that (e.g. ``function f(x:a) -> a`` won't resolve the second ``a`` properly, I think). We could resolve this by messing with scoping in name and type resolution, but we can also just already introduce proper syntax for declaring fixed free types (i.e. ``f<a>`` - we can still haggle over iterating the syntax later). That'd involve treating ``<a>`` as declaration with a proper scope in reference resolution, s.t. the uses in the signature resolve to that declaration. (Obviously, we would need to allow ``<a, b>`` and ``<a:SomeClass, b:SomeOtherClass>``, etc.)

So initial scope:
- Introduce a new ``variant`` of ``Type`` for fixed free type variables.
- Implement unification on them.
- Fix types after type-inference on function.
- Replace ``polymorphicInstance`` on importing types of already type-checked functions with a generalization step.
- Treat explicit type variables in function signatures and type class declarations as fixed free types instead of generic type variables. Optionally, adjust syntax and scoping for them in function signatures.

### PR DESCRIPTION
**Pull request by cameel**
~Depends on #14655.~ Merged.
Fixes #14570.

This PR introduces the new `forall` quantifier, which can be used to declare type variables used by a function.

Currently it's usable only on free functions. We may want to consider using it for types and instantiations as well, but those already allow declaring variables and I did not touch their syntax. We may also want to allow it inside blocks to get let-polymorphism, but that's also out of scope of this PR.

Having the quantifier, we can start requiring declarations for all type variables. Thanks to this it is now possible to refer to existing type variables and also use type variables in function return types.

### CODE REVIEW

File: test/libsolidity/syntaxTests/experimental/parsing/forall_free_function_with_sorts.sol [hunk 1]
```diff
@@ -0,0 +1,16 @@
+pragma experimental solidity;
+
+class Self: Class1 {}
+class Self: Class2 {}
+
+forall (A: (Class1, Class2), B: Class1)
```
Comment (cameel) [inline test/libsolidity/syntaxTests/experimental/parsing/forall_free_function_with_sorts.sol:6]: For now I kept the parentheses and did not use the dot as a separator. It was just easier to implement that way. I see that the parser has hacked-in support for accepting single parameters without parentheses. I could extend that but I think that what we have here might be good enough until we start working on syntax.

Comment (ekpyron) [reply test/libsolidity/syntaxTests/experimental/parsing/forall_free_function_with_sorts.sol:6]: Yeah, looks a bit weird, but fine for now :-).

File: libsolidity/analysis/NameAndTypeResolver.cpp [hunk 2]
```diff
@@ -703,6 +703,26 @@ void DeclarationRegistrationHelper::registerDeclaration(Declaration& _declaratio
...
+			!dynamic_cast<ForStatement const*>(quantifier->scope())
+		);
+		registerDeclaration(*m_scopes.at(quantifier->scope()), _declaration, nullptr, nullptr, false /* inactive */, m_errorReporter);
+
+		// TMP: Should I adjust the scope to match visibility? Here or in the Scoper?
+		//_declaration.annotation().scope = quantifier->scope();
+
+		solAssert(_declaration.annotation().contract == m_currentContract, "");
+		return;
+	}
```
Comment (cameel) [inline libsolidity/analysis/NameAndTypeResolver.cpp]: I hacked this in here for `ForAllQuantifier`, but seeing how other things related to scopes were done in #9586, I suspect we'd want to have a more general mechanism for things like this?

The problem I'm dealing with here is that `ForAllQuantifier` creates a new scope that contains its type variables and the function, which makes them invisible outside of it. For type variables it's exactly what we want but the function must be exposed to be callable.

One way to generalize this would be to make it possible to mark `Declaration`s as visible outside of their scope. I actually thought that `Declaration::isVisibleAsUnqualifiedName()` could already be used for that, but apparently it's just for hiding qualified things that are already in your scope.

Also, still not sure if the `scope` annotation not matching the scope where the declaration is registered can mess with things.

File: libsolidity/analysis/NameAndTypeResolver.cpp [hunk 3]
```diff
@@ -703,6 +703,26 @@ void DeclarationRegistrationHelper::registerDeclaration(Declaration& _declaratio
...
+		solAssert(quantifier->scope());
+		solAssert(
+			// forall quantifiers cannot be used in block scope so the declaration is always active.
+			!dynamic_cast<Block const*>(quantifier->scope()) &&
+			!dynamic_cast<ForStatement const*>(quantifier->scope())
+		);
+		registerDeclaration(*m_scopes.at(quantifier->scope()), _declaration, nullptr, nullptr, false /* inactive */, m_errorReporter);
+
+		// TMP: Should I adjust the scope to match visibility? Here or in the Scoper?
+		//_declaration.annotation().scope = quantifier->scope();
```
Comment (ekpyron) [inline libsolidity/analysis/NameAndTypeResolver.cpp]: Since this is experimental-only, we can also wait and see if and when this will cause problems.
In general rather nice that this approach doesn't seem to take *that* much hacking in the name and type resolver and all is localized here!

File: test/libsolidity/syntaxTests/experimental/parsing/forall_free_function_with_sorts.sol [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+function f(a: A: Class1, b: B: Class1) {}
+
+forall A: Class1
+function g(a: A) {}
+// ====
+// EVMVersion: >=constantinople
+// compileViaYul: true
+// ----
+// Warning 2264: (0-29): Experimental features are turned on. Do not use experimental features on live deployments.
+// TypeError 2217: (88-94): Attempt to type identifier referring to unexpected node.
```
Comment (cameel) [inline test/libsolidity/syntaxTests/experimental/parsing/forall_free_function_with_sorts.sol]: This bit is still TODO. Sorts for type variables pass fine through parsing but are rejected by type inference.

File: libsolidity/experimental/analysis/TypeInference.cpp [hunk 5]
```diff
@@ -128,6 +128,12 @@ bool TypeInference::analyze(SourceUnit const& _sourceUnit)
 	return !m_errorReporter.hasErrors();
 }
 
+bool TypeInference::visit(ForAllQuantifier const& /*_forAllQuantifier*/)
+{
+	// Nothing to do here. All the magic happens when we visit the parameter list.
```
Comment (ekpyron) [inline libsolidity/experimental/analysis/TypeInference.cpp]: That doesn't seem quite right to me - wouldn't it still need to switch to type context? I.e. won't it parse the variable declaration in the parameter list as *terms* now instead of as types?

File: libsolidity/ast/AST.h [hunk 6]
```diff
@@ -2625,6 +2625,37 @@ class Builtin: public Expression
 	SourceLocation m_nameParameterLocation;
 };
 
+class ForAllQuantifier: public ASTNode, public Scopable, public ScopeOpener
```
Comment (cameel) [inline libsolidity/ast/AST.h:2629]: Should I call it like this or would something like `UniversalQuantifier` be a better name? Any other suggestions?

File: libsolidity/experimental/analysis/TypeInference.cpp (hunk 5 above)
Comment (ekpyron) [reply libsolidity/experimental/analysis/TypeInference.cpp]: In effect the main difference would be if ``forall a:C`` treats ``C`` as type or as class...

Comment (cameel) [reply libsolidity/experimental/analysis/TypeInference.cpp]: Yeah, you're probably right. Might be why I'm getting this error: https://github.com/ethereum/solidity/pull/14668#discussion_r1383516823.

File: test/libsolidity/syntaxTests/experimental/parsing/forall_free_function_no_parentheses_single_type_var.sol [hunk 7]
```diff
@@ -0,0 +1,34 @@
+pragma experimental solidity;
+
+// The parser currently allows unary parameter lists without parentheses
```
Comment (ekpyron) [inline test/libsolidity/syntaxTests/experimental/parsing/forall_free_function_no_parentheses_single_type_var.sol]: And it has a fat TODO about that - I'd not change that for the time being, but is there really any merit in adding tests for this stuff at this point :-D? But yeah, just saying - fine with keeping it here :-).

File: libsolidity/analysis/NameAndTypeResolver.cpp (hunk 2 above)
Comment (ekpyron) [reply libsolidity/analysis/NameAndTypeResolver.cpp]: I mean, this won't happen non-experimentally - and in the final version of experimental solidity, I don't see us still using the name and type resolver of legacy solidity - that's just a temporary hack to not have to bother with doing it properly for now. So since this hack seems to work and is very localized and very clearly only applies to experimental solidity, I'd say it's fine.

File: libsolidity/experimental/analysis/TypeInference.cpp (hunk 5 above)
Comment (ekpyron) [reply libsolidity/experimental/analysis/TypeInference.cpp]: Yep, I was just about to comment that there :-).

File: libsolidity/ast/AST.h (hunk 6 above)
Comment (ekpyron) [reply libsolidity/ast/AST.h:2629]: Fine as is.

File: libsolidity/analysis/NameAndTypeResolver.cpp (hunk 3 above)
Comment (cameel) [reply libsolidity/analysis/NameAndTypeResolver.cpp]: Yeah. I was expecting way worse.

Though this could grow if we allow using `forall` with things other than just functions. 

Actually, would you use it for types or instantiations as well? Or is the current way of declaring type variables there better? I initially thought that it would fit there, but after thinking about it, it actually just makes things more verbose. Especially since, unlike with functions, repeating a type variable is not really possible in those cases:

```solidity
type T(A, B) = (A, B)

instantiation T(A: C1, B: C2): C {}
```

vs

```solidity
forall (A, B)
type T(A, B) = (A, B)

forall (A: C1, B: C2)
instantiation T(A, B): C {}
```

or would something like this even make sense?

```solidity
forall (A, B)
type T = (A, B)

forall (A: C1, B: C2)
instantiation T: C {}
```

File: test/libsolidity/syntaxTests/experimental/parsing/forall_free_function_no_parentheses_single_type_var.sol (hunk 7 above)
Comment (cameel) [reply test/libsolidity/syntaxTests/experimental/parsing/forall_free_function_no_parentheses_single_type_var.sol]: Well, this was initially a negative test checking that this generates an error and, to my surprise, it didn't - but confusingly only for a single argument. Then I found the TODO :)

I guess you have a point. Maybe I'm better off removing it.

File: libsolidity/ast/AST.h (hunk 6 above)
Comment (cameel) [inline libsolidity/ast/AST.h:2629]: One more problem here is NatSpec. Not sure if `ForAllQuantifier` not being `StructurallyDocumented` changes this, but I assume that if you put NatSpec over it, it will not be applied to function. You'd have to put it between `forall` and the function, which would look weird and be unintuitive to users.

File: libsolidity/analysis/NameAndTypeResolver.cpp (hunk 3 above)
Comment (cameel) [reply libsolidity/analysis/NameAndTypeResolver.cpp]: > Since this is experimental-only, we can also wait and see if and when this will cause problems.

After taking a closer look at how the registration and `scope()` annotation work, I think this will not cause any problems now, but it may make the interpretation of `scope()` on functions a bit error prone.

The *declaration->scope* mapping from annotations seems to be the used by name resolver only to construct the reverse *scope->declaration* mapping and then it's that reverse mapping that's used for name lookups. So this will work just fine.

The issue with code accessing `scope()` directly, on the other hand, is that it will have to explicitly account for the quantifier. Right now I allowed the quantifier only on free functions and I have not seen any checks that would be affected, but when we reintroduce contracts, this will break a lot of checks whether a function belongs to a contract. Currently they assume that the contract must be the function's scope.

I made a quick attempt to change quantified function scope in `Scoper`, but this breaks `DeclarationRegistrationHelper`'s assumptions about tree traversal - it assumes that scopes follow the nesting of `ScopeOpener`s. This can be adjusted, but I did not want the special-casing to spread to multiple places so I reverted that change in the end.

One final thing to keep in mind is also that the scope is visible in the AST output. So it may also break external tools that make assumptions about how functions are nested.

File: libsolidity/experimental/analysis/TypeInference.cpp (hunk 5 above)
Comment (cameel) [reply libsolidity/experimental/analysis/TypeInference.cpp]: Fixed.

File: test/libsolidity/syntaxTests/experimental/parsing/forall_free_function_no_parentheses_single_type_var.sol (hunk 7 above)
Comment (cameel) [reply test/libsolidity/syntaxTests/experimental/parsing/forall_free_function_no_parentheses_single_type_var.sol]: Tests removed.

**Comment by cameel** [comment]
I think that this PR with dependencies now covers everything mentioned in #14570.

File: libsolidity/ast/AST.h (hunk 6 above)
Comment (ekpyron) [reply libsolidity/ast/AST.h:2629]: We don't need to deal with Natspec for this now - the question is rather how do we make sure not to forget about this eventually :-). We should probably start a list of flukes to go over in the future including this... we could then clear out that list regularly (assuming that a lot will change in between and things may become obsolete - but might still be a bit of peace of mind to just *have* such a list to drop stuff like this into)

File: libsolidity/parsing/Parser.h [hunk 8]
```diff
@@ -102,7 +102,8 @@ class Parser: public langutil::ParserBase
 	ASTPointer<OverrideSpecifier> parseOverrideSpecifier();
 	StateMutability parseStateMutability();
 	FunctionHeaderParserResult parseFunctionHeader(bool _isStateVariable);
-	ASTPointer<ASTNode> parseFunctionDefinition(bool _freeFunction = false, bool _allowBody = true);
+	ASTPointer<ForAllQuantifier> parseQuantifiedFunctionDefinition();
+	ASTPointer<FunctionDefinition> parseFunctionDefinition(bool _freeFunction = false, bool _allowBody = true);
```
Comment (ekpyron) [inline libsolidity/parsing/Parser.h:106]: This is quite a sneaky change in here :-). Yes, it does have its own commit... but it still doesn't really belong in this PR and such :-)... but ok, I won't complain about it this time, let's just keep it :-).

File: libsolidity/ast/AST.h (hunk 6 above)
Comment (cameel) [reply libsolidity/ast/AST.h:2629]: Added a TODO comment for now.

As for the list - do you think an issue a good place for it or would you just have a HackMD note instead?

Comment (ekpyron) [reply libsolidity/ast/AST.h:2629]: Rather a hackmd note, I'd say

File: libsolidity/parsing/Parser.h (hunk 8 above)
Comment (cameel) [reply libsolidity/parsing/Parser.h:106]: It's related. Without this I'd have to cast the result of `parseFunctionDefinition()` to `FunctionDefinition` in the code I added. I could do that, but come on, the function is called `parseFunctionDefinition()` and it does not make sense for it to return anything other than a `FunctionDefinition`.

Comment (ekpyron) [reply libsolidity/parsing/Parser.h:106]: Ok, fair enough. Not that in the long run the forall node shouldn't probably take a more general type than a function definition, but fine to keep both as is now.

File: libsolidity/ast/AST.h (hunk 6 above)
Comment (cameel) [reply libsolidity/ast/AST.h:2629]: [Experimental Solidity quirks to deal with eventually](https://notes.ethereum.org/TLYlbuvqSGKIzRv6F4Mxow).

**Comment by cameel** [comment]
Some jobs are failing, but this does not seem related to the PR. Looks like #14839 might not have fixed everything.

**Review by ekpyron — APPROVED** [review]

**Comment by nikola-matic** [comment]
> Some jobs are failing, but this does not seem related to the PR. Looks like #14839 might not have fixed everything.

Well, this is lame; looks like the image used in this PR has a different version of Python

```
circleci@packer-651dab20-1061-caa7-7f59-31ffa2e7d7e2  ~
$ which python
/c/Python311/python
```

It's 311 instead of 312, so obviously can't find the binary to symlink since it doesn't exist.

**Comment by cameel** [comment]
Hmmm... so I guess CircleCI has multiple different versions of the Windows image and they assign them randomly to CI runs? This would also explain why miniconda just suddenly disappeared without us touching the config.

If that's the case, we might be best off just doing #14847 right away to get rid of the hard-coded path.

**Comment by nikola-matic** [comment]
> Hmmm... so I guess CircleCI has multiple different versions of the Windows image and they assign them randomly to CI runs? This would also explain why miniconda just suddenly disappeared without us touching the config.
> 
> If that's the case, we might be best off just doing #14847 right away to get rid of the hard-coded path.

Agreed.
